include/seahorn/Expr/ExprOpBinder.hh

```diff
@@ -126,14 +125,62 @@ struct BINDER {
 };
 } // namespace bind
 
+namespace typeCheck {
+namespace bindType {
+
+template <typename T> static inline bool checkBody(Expr exp, TypeChecker &tc) {
+  Expr body = exp->last();
+
+  if (!(exp->arity() >= 2 && correctTypeAny<T>(body, tc)))
+    return false;
+
+  ExprSet bodyEndExps = tc.getEndExps(body);
+  ExprSet boundVars(exp->args_begin(), exp->args_end() - 1);
+
+  auto contains = [&boundVars](Expr exp) -> bool {
+    return boundVars.count(exp);
+  };
+
+  return std::all_of(bodyEndExps.begin(), bodyEndExps.end(), contains);
+}
+
+struct Lambda {
```

Comment:
lambda's are functional type. For example, `lambda x :: x + 1` has type: `typeOf(x) -> typeOf(x + 1)`
where `typeOf` is the type of appropriate expression. Internally, lambda arguments are represented with bound variables. So the expression above looks like: `lambda :: B0 + 1`, where `B0` is the only bound variable and is therefore the argument. You can infer the type for arguments from the lambda operator itself (it keep track of how many variables is bounds and their sorts), you can infer the type of the body from type checking the body recursively.